### PR TITLE
Derive classification profiles

### PR DESCRIPTION
**Pull request by roshankern**
This PR is ready for review!

In this PR, classification profiles are derived from the single cell classifications. These classification profiles are simply the single cell classifications averaged across perturbation.

Only 4b.derive-classification-profiles.py and the readme changes need to be reviewed as the other files should have already been reviewed in other PRs (but another review is always welcome!). There are about 190 lines that need to be reviewed.

### CODE REVIEW

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import classification_utils
+
+
+# ### Load Cell Health Profile Labels
+# 
+
+# In[2]:
+
+
+cell_health_hash = "30ea5de393eb9cfc10b575582aa9f0f857b44c59"
```
Comment (d33bs) [inline 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py]: Consider providing a comment above this line to indicate the relative datetime related to this hash.

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+cell_health_labels
+
+
+# ### Create Classification Profiles
+# 
+
+# In[3]:
+
+
+# which cell lines correspond to whice plates (from Cell Health IDR metadata)
```
Comment (d33bs) [inline 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py:40]: Consider providing a link to the webpage or file with the IDR metadata here.

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+# In[4]:
+
+
+# paths to set (data is loaded from/saved to external hard drive)
+cell_health_plate_classifications = pathlib.Path(
+    "/media/roshankern/63af2010-c376-459e-a56e-576b170133b6/data/cell-health-plate-classifications"
+)
+classification_profiles_save_dir = pathlib.Path(
+    "/media/roshankern/63af2010-c376-459e-a56e-576b170133b6/data/cell-health-plate-classification-profiles"
```
Comment (d33bs) [inline 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py]: Consider using a common variable for the base directory used in both of these paths to reduce duplication.

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# ### Import Libraries
+# 
+
+# In[1]:
+
+
+import importlib
```
Comment (d33bs) [inline 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py]: Consider removing this import if it is not used.

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+import pathlib
+import sys
+
+import pandas as pd
+import seaborn as sns
+import matplotlib.pyplot as plt
```
Comment (d33bs) [inline 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py]: Consider removing these imports if they are not used.

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    classification_profiles_save_path.parent.mkdir(exist_ok=True, parents=True)
+    classification_profiles.to_csv(classification_profiles_save_path, sep="\t", index=False)
+
+# derive single class model classification profiles
+print("Deriving single class model classification profiles")
+for phenotypic_class_dir in SCM_classifications.iterdir():
+    print(f"Deriving classification profiles for {phenotypic_class_dir.name} models")
+    for model_classifications_dir in phenotypic_class_dir.iterdir():
+
+        # get information about the current model
```
Comment (d33bs) [inline 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py:107]: Consider providing more context about the below model information captured from the directory names, maybe in the form of an example or an f-string which labels the name components accessed from splitting.

File: 4.classify-single-cell-phenotypes/classification_utils.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+import numpy as np
+import pandas as pd
+from sklearn.linear_model import LogisticRegression
+
+
+def get_probas_dataframe(
+    plate_features: pd.DataFrame, model: LogisticRegression, feature_type: str
+) -> pd.DataFrame:
+    """
+    Get probabilites for plate features from a phenotypic classification model
```
Comment (d33bs) [inline 4.classify-single-cell-phenotypes/classification_utils.py]:
```suggestion
    Get probabilities for plate features from a phenotypic classification model
```

File: 4.classify-single-cell-phenotypes/classification_utils.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
+"""
+utils for classifying cells from Cell Health Data
+"""
+
+import pathlib
+import numpy as np
```
Comment (d33bs) [inline 4.classify-single-cell-phenotypes/classification_utils.py]: Consider removing this import if it is not used.

File: 4.classify-single-cell-phenotypes/classification_utils.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+import pandas as pd
+from sklearn.linear_model import LogisticRegression
+
+
+def get_probas_dataframe(
+    plate_features: pd.DataFrame, model: LogisticRegression, feature_type: str
```
Comment (d33bs) [inline 4.classify-single-cell-phenotypes/classification_utils.py]: In context with the docstring below, consider using a [`typing.Literal` type hint](https://docs.python.org/3/library/typing.html#typing.Literal) for the `feature_type` parameter to help indicate the expectation of specific string values.

File: 4.classify-single-cell-phenotypes/classification_utils.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+    Returns
+    -------
+    pd.DataFrame
+        dataframe with single-cell probabilities for classes from given model
+    """
+
+    # determine which feature columns should be loaded depending on feature type
+    # if there is no "and" we can use feature type as prefix
+    if "and" not in feature_type:
```
Comment (d33bs) [inline 4.classify-single-cell-phenotypes/classification_utils.py]: Consider using a specific check for the literal string value which includes "and" (there appears to be only one circumstance where this occurs as-is).
```suggestion
    if feature_type != "CP_and_DP":
```

File: 4.classify-single-cell-phenotypes/classification_utils.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    probas_dataframe = pd.DataFrame(
+        model.predict_proba(single_cell_features),
+        columns=model.classes_,
+    ).reset_index(drop=True)
+
+    return probas_dataframe
+
+
+def create_classification_profiles(
+    plate_classifications_dir: pathlib.Path, cell_line_plates: dict
```
Comment (d33bs) [inline 4.classify-single-cell-phenotypes/classification_utils.py:61]: Consider using [`typing.Dict`](https://docs.python.org/3/library/typing.html#typing.Dict) here with subscripts based on the expected key and value types for `cell_line_plates` (note: Python 3.9 allows subscripting the dictionary type hint but I noticed this environment uses 3.8).

**Review by d33bs — APPROVED** [review]
Nice work! I thought things looked good and only had minor comments. Please let me know if you have any questions.

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py (hunk 1 above)
Comment (MattsonCam) [reply 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py]: Maybe you are already doing this, but you could consider printing this out after retrieval from a hashmap.

File: 4.classify-single-cell-phenotypes/classification_utils.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+    # determine which feature columns should be loaded depending on feature type
+    # if there is no "and" we can use feature type as prefix
+    if "and" not in feature_type:
+        cols_to_load = [
+            col
+            for col in plate_features.columns.to_list()
+            if f"{feature_type}__" in col
+        ]
+    # if there is an "and" we should all features (which all have "P__" prefix)
```
Comment (MattsonCam) [inline 4.classify-single-cell-phenotypes/classification_utils.py]: Consider saving columns with more descriptive names, or commenting on the distinction.

File: 4.classify-single-cell-phenotypes/classification_utils.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    cell_line_plates : dict
+        cell line names, each with a list of plate names that are correlated to their respective cell line
+
+    Returns
+    -------
+    pd.DataFrame
+        classification profiles dataframe
+    """
+
+    cell_line_classification_profiles = []
```
Comment (MattsonCam) [inline 4.classify-single-cell-phenotypes/classification_utils.py:80]: Consider performing more of these steps outside of the for loops or by adding more modularity in another way. 

Good documentation though!

**Review by MattsonCam — APPROVED** [review]
Good job! This helps me understand the information we are drawing from the data for classifications. Just a few suggestions, please feel free to let me know if you have questions!

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.ipynb [hunk 14]
```diff
@@ -0,0 +1,607 @@
+{
```
Comment (MattsonCam) [inline 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.ipynb:1]: <p>May consider adding a markdown objective description at the top</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/WayScience/cell-health-data/pull/20/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='0'/>

Comment (MattsonCam) [inline 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.ipynb:1]: <p>Consider storing this data, for reference.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/WayScience/cell-health-data/pull/20/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='5'/>

Comment (MattsonCam) [inline 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    cell_health_plate_classifications = pathlib.Path(
</pre><p>If this data is not confidential or not too large, consider allowing open access</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/WayScience/cell-health-data/pull/20/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='6'/>

Comment (MattsonCam) [inline 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #18.    print("Deriving multi class model classification profiles")
</pre><p>I like the indicator line here</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/WayScience/cell-health-data/pull/20/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='6'/>

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py (hunk 1 above)
Comment (roshankern) [reply 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.py]: > Maybe you are already doing this, but you could consider printing this out after retrieval from a hashmap.

Yep, already doing this!

File: 4.classify-single-cell-phenotypes/classification_utils.py (hunk 11 above)
Comment (roshankern) [reply 4.classify-single-cell-phenotypes/classification_utils.py:61]: Given that Python 3.9 does not have this feature, I think it is not worth attempting to implement this change.

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.ipynb (hunk 14 above)
Comment (roshankern) [reply 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.ipynb:1]: I think that this is not necessary here given the info in the readme file.

Comment (roshankern) [reply 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.ipynb:1]: The plan is to upload this data to Zenodo (open access data storage) in the near future! I will include the info for accessing this data as soon as possible.
